### PR TITLE
dynamic judgment PASCAL

### PR DESCRIPTION
**Pull request by yajiedesign**
dependent https://github.com/dmlc/mshadow/pull/237

### CODE REVIEW

**Comment by piiswrong** [comment]
@DickJC123

File: src/operator/deconvolution.cu [hunk 1]
```diff
@@ -32,17 +32,21 @@ Operator* CreateOp<gpu>(DeconvolutionParam param, int dtype,
...
-  int desired_forward_compute_type =
-    (dtype == mshadow::kFloat16) ? mshadow::kFloat32 : dtype;
-  int desired_backward_compute_type =
-    (dtype == mshadow::kFloat16) ? mshadow::kFloat32 : dtype;
-#endif  // MSHADOW_USE_PASCAL == 1
+
+
+  #if CUDNN_MAJOR >= 6
+  cudaDeviceProp prop;
+  cudaGetDeviceProperties(&prop, ctx.dev_id);
```
Comment (piiswrong) [inline src/operator/deconvolution.cu]: this is already in the stream right?

File: src/operator/convolution.cu [hunk 2]
```diff
@@ -33,17 +33,21 @@ Operator* CreateOp<gpu>(ConvolutionParam param, int dtype,
...
-#else
-  // pseudo fp16
-  int desired_forward_compute_type =
-    (dtype == mshadow::kFloat16) ? mshadow::kFloat32 : dtype;
-  int desired_backward_compute_type =
-    (dtype == mshadow::kFloat16) ? mshadow::kFloat32 : dtype;
-#endif  // MSHADOW_USE_PASCAL == 1
+
+
+#if CUDNN_MAJOR >= 6
```
Comment (piiswrong) [inline src/operator/convolution.cu]: @DickJC123 could you review?

Comment (DickJC123) [inline src/operator/convolution.cu]: One issue is whether a published model runs consistently across all users.  For example, a user with a MAXWELL brick could post a model with float16 data and advertise certain behavior.  Someone who attempts to run the model on a PASCAL brick might see vastly different behavior, not realizing that he was running true fp16, where the model author was running pseudo fp16.  Right now the MSHADOW_USE_PASCAL=0 flag enforces pseudo fp16 for all.  After compiling with MSHADOW_USE_PASCAL=1, true fp16 is attempted and MAXWELL owners will get a warning that the desired true fp16 behavior is not achievable, and the system will fall back to pseudo fp16.  There's no mechanism currently to quiet this warning, which might get annoying.

Now I don't particularly like the MSHADOW_USE_PASCAL compiled-in flag, although I was under the impression that you wanted us to use it.  I would like to see the ability to set a global policy for how to handle operations on float16 tensors with the following choices:
- platform attempts true fp16, throws exception if not achievable
- platform attempts true fp16, falls back to pseudo fp16 (i.e. 32-bit compute) with warning
- platform attempts true fp16, silently falls back to pseudo fp16.

And while we're on the subject, I like the idea of really fine-tuning the compute precision on a per-convolution basis, both to achieve the highest performance, but also to really fix the behavior where needed to preserve functionality across GPUs.  I'd vote for adding Convolution Symbol parameters that might be used as in: cudnn_algo_fwd_prec=float32, cudnn_algo_bwd_prec=float16.

Comment (yajiedesign) [inline src/operator/convolution.cu]: @DickJC123 i have some question. i not see "fall back to pseudo" code with  MSHADOW_USE_PASCAL=1.the `CuDNNConvolutionOp<DType>::Supports` not check device.

**Comment by mli** [comment]
I'm going to close the PR since there is no activity for more than 2 months. But please feel free to re-open it and update with the master.

**Pull request closed without merging**